### PR TITLE
Add metrics to monitor size of ongoing raft snapshots

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: ref #7777

Problem Summary:

1. change type_apply_snapshot_predecode and type_apply_snapshot_predecode_sst2dt from order 10 to order 15.
2. add a metric to observe all ongoing snapshot size on disk.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Common/TiFlashMetrics.h [hunk 1]
```diff
@@ -410,9 +410,9 @@ namespace DB
     M(tiflash_raft_raft_log_gap_count,                                                                                              \
       "Bucketed histogram raft index gap between applied and truncated index",                                                      \
       Histogram,                                                                                                                    \
-      F(type_applied_index, {{"type", "applied_index"}}, EqualWidthBuckets{0, 100, 10}),                                            \
+      F(type_applied_index, {{"type", "applied_index"}}, EqualWidthBuckets{0, 100, 15}),                                            \
```
Comment (CalvinNeo) [inline dbms/src/Common/TiFlashMetrics.h:413]: I find in raftstore-v2 scene, raft log gap may be more than 1000

File: dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.cpp [hunk 2]
```diff
@@ -229,12 +237,15 @@ void SSTFilesToBlockInputStream::loadCFDataFromSST(
             region->insert(cf, TiKVKey(key.data, key.len), TiKVValue(value.data, value.len), DupCheck::AllowSame);
             reader->next();
             (*p_process_keys) += 1;
+            (*p_process_keys_bytes) += (key.len + value.len);
         }
         LOG_DEBUG(
             log,
-            "Done loading all kvpairs from [CF={}] [offset={}] [write_cf_offset={}] [region_id={}] [snapshot_index={}]",
+            "Done loading all kvpairs from [CF={}] [offset={}] [proc_size={}] [write_cf_offset={}] [region_id={}] "
+            "[snapshot_index={}]",
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.cpp:245]:
```suggestion
            "Done loading all kvpairs, CF={} offset={} processed_size={} write_cf_offset={} region_id={} "
            "snapshot_index={}",
```
I think `processed_size` is better than `proc_size`

File: dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.cpp [hunk 3]
```diff
@@ -250,10 +261,11 @@ void SSTFilesToBlockInputStream::loadCFDataFromSST(
         {
             LOG_DEBUG(
                 log,
-                "Done loading from [CF={}] [offset={}] [write_cf_offset={}] [last_loaded_rowkey={}] "
+                "Done loading from [CF={}] [offset={}] [proc_size={}] [write_cf_offset={}] [last_loaded_rowkey={}] "
                 "[rowkey_to_be_included={}] [region_id={}] [snapshot_index={}]",
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.cpp:265]:
```suggestion
                "Done loading, CF={} offset={} processed_size={} write_cf_offset={} last_loaded_rowkey={} "
                "rowkey_to_be_included={} region_id={} snapshot_index={}",
```
ditto

File: dbms/src/Storages/KVStore/MultiRaft/ApplySnapshot.cpp [hunk 4]
```diff
@@ -437,7 +439,11 @@ std::vector<DM::ExternalDTFileInfo> KVStore::preHandleSSTsToDTFiles(
                     index);
                 stream->cancel();
             }
-            generated_ingest_ids = stream->outputFiles();
+            prehandle_result.ingest_ids = stream->outputFiles();
+            prehandle_result.stats = PrehandleResult::Stats{
+                sst_stream->getProcessKeys().total_bytes(),
+                stream->getTotalCommittedBytes(),
+                stream->getTotalBytesOnDisk()};
```
Comment (JaySon-Huang) [inline dbms/src/Storages/KVStore/MultiRaft/ApplySnapshot.cpp:446]: Seems it should be 
```suggestion
            prehandle_result.stats = PrehandleResult::Stats{
                sst_stream->getProcessKeys().total_bytes(),
                stream->getTotalBytesOnDisk(),
                stream->getTotalCommittedBytes(),
                };
```

Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/MultiRaft/ApplySnapshot.cpp:446]: wtf, i must be out of my mind...

**Review by hongyunyan — APPROVED** [review]

**Comment by JaySon-Huang** [comment]
This is not fixed yet https://github.com/pingcap/tiflash/pull/8095#discussion_r1325985075

File: dbms/src/Common/TiFlashMetrics.h [hunk 5]
```diff
@@ -440,6 +440,12 @@ namespace DB
       "Bucketed histogram entry size",                                                                                              \
       Histogram,                                                                                                                    \
       F(type_normal, {{"type", "normal"}}, ExpBuckets{1, 2, 13}))                                                                   \
+    M(tiflash_raft_ongoing_snapshot_total_size,                                                                                     \
```
Comment (JaySon-Huang) [inline dbms/src/Common/TiFlashMetrics.h:443]:
```suggestion
    M(tiflash_raft_ongoing_snapshot_total_bytes,                                                                                     \
```

File: metrics/grafana/tiflash_summary.json [hunk 6]
```diff
@@ -11359,6 +11359,105 @@
...
+          "points": false,
+          "renderer": "flot",
+          "seriesOverrides": [],
+          "spaceLength": 10,
+          "stack": false,
+          "steppedLine": false,
+          "targets": [
+            {
+              "exemplar": true,
+              "expr": "sum(rate(tiflash_raft_ongoing_snapshot_total_size{k8s_cluster=\"$k8s_cluster\", tidb_cluster=\"$tidb_cluster\", instance=~\"$instance\"}[$__rate_interval])) by (type)",
```
Comment (JaySon-Huang) [inline metrics/grafana/tiflash_summary.json:11411]: we should also update this

File: dbms/src/Debug/dbgTools.cpp [hunk 7]
```diff
@@ -779,8 +779,10 @@ void handleApplySnapshot(
     TMTContext & tmt)
 {
     auto new_region = kvstore.genRegionPtr(std::move(region), peer_id, index, term);
-    auto external_files = kvstore.preHandleSnapshotToFiles(new_region, snaps, index, term, deadline_index, tmt);
-    kvstore.applyPreHandledSnapshot(RegionPtrWithSnapshotFiles{new_region, std::move(external_files)}, tmt);
+    auto prehandle_result = kvstore.preHandleSnapshotToFiles(new_region, snaps, index, term, deadline_index, tmt);
+    kvstore.applyPreHandledSnapshot(
+        RegionPtrWithSnapshotFiles{new_region, std::move(prehandle_result.ingest_ids)},
```
Comment (CalvinNeo) [inline dbms/src/Debug/dbgTools.cpp:784]: The rest part of prehandle_result is useless here

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**